proposals/p2665.md

```diff
@@ -0,0 +1,209 @@
+an `if` body to be on the same line in order to avoid semicolon insertion.
+
+## Proposal
+
+As described in the abstract, Carbon will require semicolons to terminate
+statements and forward declarations.
+
+Examples with a semicolon include:
+
+-   Most statements, such as `Foo();` and `x = x + 2;`.
+-   `var` statements and declarations, such as `var x: i32 = 0;`
+-   Forward declarations, such as `class C;` or `fn Foo();`.
+
+Examples with a close curly brace include:
+
+-   Statement grammars that terminate with a curly brace, such as
+    `if ( ... ) { ... }` or `match ( ... ) { ... }`.
+-   Declarations that include a definition, such as `class C { ... }` or
+    `fn Foo() { ... }`.
+    -   This is partly in contrast with C++, which would requires a semicolon in
+        `class C { ... };`.
+
+Carbon's current design has been written assuming the above; this is making
+requiring semicolons an explicit decision.
+
+## Rationale
+
+-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
+    -   We expect it to be easier to write tools that parse and operate on
+        source code if semicolons are required.
+-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
+    -   Requiring semicolons leaves open the most evolutionary paths; any
+        optional semicolon approach means the design would need to be more
+        thoughtful about handling ambiguities.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Semicolons are a
+        [visual aid](/docs/project/principles/low_context_sensitivity.md#visual-aids)
+        that reinforces statement termination, even though they might be viewed
+        as a nuisance to write or visually unnecessary for some developers.
+        -   Carbon weighs readability more heavily because of the expectation
+            that code will be read more often.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   The use of semicolons is expected to improve familiarity for C++
+        developers, even for developers who might prefer optional semicolons.
+
+## Alternatives considered
+
+### Optional semicolons
+
+Semicolons could be made optional. This would most likely be with an approach
+similar to Python, based mainly on newlines.
+
+Advantages:
+
+-   Languages with optional semicolons are very popular. Python is either the
+    most, or the 2nd most, widely used programming language by most measures
+    ([1](https://pypl.github.io/PYPL.html)
+    [2](https://octoverse.github.com/2022/top-programming-languages)
+    [3](https://www.tiobe.com/tiobe-index/)).
+-   Echoes the direction of evolution in other languages.
+    -   Swift and Kotlin are examples of recently developed languages that make
+        semicolons optional in good ways.
```

Comment:
Can we say something more precise than "good" here? Is this "in ways that work well for developers in practice"?

Swift developers seem to like the Swift rule, and it seems hard to find any realistic examples where it does the wrong thing (even though it's not totally clear what rule Swift is using), but I didn't have trouble finding forum posts of Kotlin developers complaining about the Kotlin rule in practice.

Reply:
Rephrased, but I'm not quickly finding the kotlin rule complaints. What are your search terms?

I did find https://stackoverflow.com/questions/49034127/kotlin-is-ignoring-expression, but the related complaint seems to be a compiler error now.
https://kotlin.compiler-explorer.com/z/5E1GreqjM